### PR TITLE
Dropout and Validation Sets

### PR DESCRIPTION
**Pull request by rochan17**
Added dropout and validation capability. 

- most changes are due to addition of black formatting
- Dropout added in the unet model, with focus on allowing it to be configured separately at each resolution. 
- added validation set and indices capability in dataset and dataloading. Now one can specify indices of the dataset to make the train/val tests if they don't want random sampling. 
- subset can also be stated if random sampling from specific subset is required (limited as of now to single regions subsets)

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -4,18 +4,25 @@
...
-        
+
+    def __init__(
+        self,
+        data_path,
+        img_size=(64, 64, 64),
+        length=5,
+        apply_symmetry=True,
+        indices=None,
+        subset=None,
```
Comment (rkurchin) [inline data.py:24]: add docstring entries for these!

File: data.py [hunk 2]
```diff
@@ -42,43 +49,68 @@ def __init__(self,data_path,img_size=(64,64,64),length=5,apply_symmetry=True):
...
+
     def __len__(self):
         return self.total_samples
 
     def __getitem__(self, idx):
-        
-        im            = self.data
-        im_dims       = np.array(self.data.shape)
+
+        im = self.data
```
Comment (rkurchin) [inline data.py:80]: get rid of this and just reference self.data below, will likely save some memory since I doubt Python is smart enough 😉

File: main.py [hunk 3]
```diff
@@ -1,99 +1,136 @@
...
 
 def main():
-    
+
     global args
-    
-    if args.n_epochs < 5*args.sample_interval:
-        args.sample_interval = args.n_epochs//6
+
+    if args.n_epochs < 5 * args.sample_interval:
```
Comment (rkurchin) [inline main.py:43]: add a comment line or two here; it's not obvious what `sample_interval` is or why it should/shouldn't be related to epoch count like this

File: main.py [hunk 4]
```diff
@@ -102,48 +139,61 @@ def main():
 class MicroData(LightningDataModule):
```
Comment (rkurchin) [inline main.py:139]: should this be in this file?

File: main.py [hunk 5]
```diff
@@ -153,143 +203,349 @@ def __init__(
...
-    parser.add_argument("--is_attn", type=str, default="(0,0,1,1)", help="Whether the block has self-attention")
-    parser.add_argument("--var_schedule", type=str, default="linear", help="Diffusion variance schedule :- linear or cosine")
-    parser.add_argument("--beta_start", type=float, default=0.0001, help="variance schedule start")
-    parser.add_argument("--beta_end", type=float, default=0.02, help="variance schedule end")
-    
+    parser.add_argument(
+        "--base_dim", type=int, default=16, help="base dimension in the UNet"
+    )
+    parser.add_argument(
+        "--sample_interval", type=int, default=100, help="interval betwen image samples"
```
Comment (rkurchin) [inline main.py:486]: typo

File: sampler.py [hunk 6]
```diff
@@ -11,136 +11,160 @@
...
+    num_timesteps=1000,
+    linear_start=0.0001,
+    linear_end=0.02,
+    cosine_s=0.01,
+):
+
+    if schedule == "linear":
+        betas = torch.linspace(linear_start, linear_end, num_timesteps)
+
+    elif schedule == "cosine":
```
Comment (rkurchin) [inline sampler.py:25]: some comments here would be good too

File: sampler.py [hunk 7]
```diff
@@ -11,136 +11,160 @@
...
-        
+
         # For denoising step
-        
+
         if inf_timesteps > diff_timesteps:
-            raise ValueError('''Number of Inference Steps > Number of Diffusion Steps \U0001F928''')
-            
+            raise ValueError(
+                """Number of Inference Steps > Number of Diffusion Steps \U0001F928"""
```
Comment (rkurchin) [inline sampler.py:72]: 🤨

File: unet_attention.py [hunk 8]
```diff
@@ -12,55 +12,69 @@
...
+        self.gamma = nn.Parameter(torch.tensor([0.0]))
 
-    def _conv(self,n_in,n_out):
-        return nn.Conv1d(n_in, n_out,1,bias=False)
+    def _conv(self, n_in, n_out):
+        return nn.Conv1d(n_in, n_out, 1, bias=False)
 
     def forward(self, x):
-        #Notation from the paper.
+        # Notation from the paper.
```
Comment (rkurchin) [inline unet_attention.py:53]: add DOI/link

File: unet_attention.py [hunk 9]
```diff
@@ -12,55 +12,69 @@
...
     """
 
-    def __init__(self, in_channels: int, out_channels: int, time_channels: int,
-                 n_groups: int = 8, dropout: float = 0.1):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        time_channels: int,
+        n_groups: int = 8,
```
Comment (rkurchin) [inline unet_attention.py:75]: missing docstring entries

File: unet_attention.py [hunk 10]
```diff
@@ -140,33 +168,47 @@ class UpBlock(nn.Module):
...
         x = self.res(x, t)
         x = self.attn(x)
         return x
-    
+
+
 class MiddleBlock(nn.Module):
     """
     ### Middle block
     It combines a `ResidualBlock`, `AttentionBlock`, followed by another `ResidualBlock`.
```
Comment (rkurchin) [inline unet_attention.py:199]: `AttentionBlock` is optional, right? Clarify/rephrase

File: unet_attention.py [hunk 11]
```diff
@@ -222,25 +268,38 @@ class UNet(nn.Module):
...
-                 n_blocks: int = 2,middle_attn=True):
+    def __init__(
+        self,
+        image_channels: int = 1,
+        n_channels: int = 64,
+        ch_mults=(1, 2, 2, 4),
+        is_attn=(False, False, False, True),
+        n_blocks: int = 2,
+        middle_attn=True,
+        dropout: float = 0.05,
```
Comment (rkurchin) [inline unet_attention.py:279]: while this is fairly obvious, do add the docstring entry since this is the "put-it-all-together" function

**Review by rkurchin — CHANGES_REQUESTED** [review]
Generally looks nice! I put a few comments, basically all about docstrings/comments. It would be good to also write some tests at some point – e.g. with a good set of tests, you wouldn't need to worry about that memory-saving fix wrt. `im`/`self.data` breaking anything, because you could just test right away (even through CI on GitHub!).
